File: Python/Partie5/Exo20.py
import threading
import time
import random
import sys

compt = 0
mutex = threading.Lock()
stop = False

def incrementeur(limite):
    global compt, stop
    while not stop:
        with mutex:
            if compt >= limite:
                stop = True
                break
            compt += 1
            print(f"Incrementeur : compt = {compt}")
        time.sleep(random.randint(1, 5))

def afficheur(limite):
    global stop
    while not stop:
        time.sleep(2)
        with mutex:
            if compt >= limite:
                stop = True
                break
            print(f"Afficheur : compt = {compt}")

if len(sys.argv) != 2:
    print("Usage: python prog.py limite")
    sys.exit(1)
limite = int(sys.argv[1])
t1 = threading.Thread(target=incrementeur, args=(limite,))
t2 = threading.Thread(target=afficheur, args=(limite,))
t1.start()
t2.start()
t1.join()
t2.join()
print(f"Valeur limite {limite} atteinte")

File: C/Partie5/Exo20.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <unistd.h>
#include <time.h>

int compt = 0;
int limite;
pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;

void *fonction_incrementeur(void *arg)
{
    while (1)
    {
        pthread_mutex_lock(&mutex);
        if (compt >= limite)
        {
            pthread_mutex_unlock(&mutex);
            break;
        }
        compt++;
        printf("Thread 1: compt = %d\n", compt);
        pthread_mutex_unlock(&mutex);

        // attend un temps aléatoire entre 1 et 5 secondes
        int duree = rand() % 5 + 1;
        sleep(duree);
    }
    return NULL;
}

void *fonction_screen(void *arg)
{
    while (1)
    {
        sleep(2); // toutes les 2 secondes

        pthread_mutex_lock(&mutex);
        if (compt >= limite)
        {
            pthread_mutex_unlock(&mutex);
            break;
        }
        printf("Thread 2: compt = %d\n", compt);
        pthread_mutex_unlock(&mutex);
    }
    return NULL;
}

int main(int argc, char *argv[])
{

    srand(time(NULL));

    pthread_t t1, t2;
    pthread_create(&t1, NULL, fonction_incrementeur, NULL);
    pthread_create(&t2, NULL, fonction_screen, NULL);

    pthread_join(t1, NULL);
    pthread_join(t2, NULL);

    printf("Valeur limite %d atteinte. Fin du programme.\n", limite);
    pthread_mutex_destroy(&mutex);
    return 0;
}